velox/exec/PrefixSortAlgorithm.h

```diff
@@ -0,0 +1,127 @@
+  PrefixSortIterator(char* prefix, const uint64_t entrySize)
+      : prefix(prefix), entrySize(entrySize) {}
+
+  PrefixSortIterator(const PrefixSortIterator& other)
+      : prefix(other.prefix), entrySize(other.entrySize) {}
+
+  char* operator*() const;
+
+  PrefixSortIterator& operator++();
+
+  PrefixSortIterator& operator--();
+
+  PrefixSortIterator operator++(int);
+
+  PrefixSortIterator operator--(int);
+
+  PrefixSortIterator operator+(const uint64_t i) const;
+
+  PrefixSortIterator operator-(const uint64_t i) const;
+
+  PrefixSortIterator& operator=(const PrefixSortIterator& other);
+
+  uint64_t operator-(const PrefixSortIterator& other) const;
+
+  bool operator<(const PrefixSortIterator& other) const;
+
+  bool operator>(const PrefixSortIterator& other) const;
+
+  bool operator>=(const PrefixSortIterator& other) const;
+
+  bool operator<=(const PrefixSortIterator& other) const;
+
+  bool operator==(const PrefixSortIterator& other) const;
+
+  bool operator!=(const PrefixSortIterator& other) const;
+
+ private:
+  char* prefix;
+  const uint64_t entrySize;
+};
+
+using SwapBufferPtr = std::unique_ptr<char[], std::default_delete<char>>;
+
+inline SwapBufferPtr makeSwapBuffer(uint64_t __n) {
+  return std::unique_ptr<char[], std::default_delete<char>>(new char[__n]());
+}
+
+struct PrefixSortContext {
+  PrefixSortContext(uint64_t entrySize, char* end)
+      : entrySize(entrySize),
+        iterSwapBufPtr(makeSwapBuffer(entrySize)),
+        iterSwapBuf(iterSwapBufPtr.get()),
+        end(end) {}
+  const uint64_t entrySize;
+  SwapBufferPtr iterSwapBufPtr;
+  char* iterSwapBuf;
+  // Point to continuous sort data end(exclusive).
+  char* end;
+
+  void iterSwap(const PrefixSortIterator& lhs, const PrefixSortIterator& rhs)
+      const;
+
+  void iterRangeSwap(
+      const PrefixSortIterator& start1,
+      const PrefixSortIterator& start2,
+      uint64_t length) const;
+};
+
+static int kSmallSort = 7;
+static int kMediumSort = 40;
+
+/// Calculate the middle of iterators a, b, and c based on the data.
+/// For example: if a, b, and c point to data 1, 2, and 3 respectively, then
+/// iterator b is returned.
+inline static PrefixSortIterator median3(
+    const PrefixSortIterator& a,
+    const PrefixSortIterator& b,
+    const PrefixSortIterator& c,
+    std::function<
+        int(const PrefixSortIterator&, const PrefixSortIterator&)> const& cmp) {
+  return cmp(a, b) < 0 ? (cmp(b, c) < 0       ? b
+                              : cmp(a, c) < 0 ? c
+                                              : a)
+                       : (cmp(b, c) > 0       ? b
+                              : cmp(a, c) > 0 ? c
+                                              : a);
+}
+
+/// Sort prefix data in range [start, end) using quick-sort.
+/// Algorithm implementation reference :
+/// 1. J.L.Bentley and M.McIlroy’s paper:
+/// “Engineering a Sort Function”
+/// 2. Presto`s quickSort implementation
+void quickSortForPrefixSort(
```

Comment:
Can this be a const method in PrefixSortContext?

PrefixSortContext::quickSort(start, end, compare)